moquette-0.16/broker/src/main/java/io/moquette/broker/Server.java

```diff
@@ -0,0 +1,400 @@
+                .newInstance(props);
+        } catch (InstantiationException | IllegalAccessException | ClassNotFoundException ex) {
+            LOG.warn("Unable to invoke constructor with {} argument. ClassName={}, interfaceName={}, cause={}, " +
+                     "errorMessage={}", constructorArgClass.getName(), className, intrface.getName(), ex.getCause(),
+                     ex.getMessage());
+            return null;
+        } catch (NoSuchMethodException | InvocationTargetException e) {
+            try {
+                LOG.info("Invoking default constructor. ClassName={}, interfaceName={}", className, intrface.getName());
+                // fallback to default constructor
+                instance = this.getClass().getClassLoader()
+                    .loadClass(className)
+                    .asSubclass(intrface)
+                    .getDeclaredConstructor().newInstance();
+            } catch (InstantiationException | IllegalAccessException | ClassNotFoundException |
+                NoSuchMethodException | InvocationTargetException ex) {
+                LOG.error("Unable to invoke default constructor. ClassName={}, interfaceName={}, cause={}, " +
+                          "errorMessage={}", className, intrface.getName(), ex.getCause(), ex.getMessage());
+                return null;
+            }
+        }
+
+        return instance;
+    }
+
+    /**
+     * Use the broker to publish a message. It's intended for embedding applications. It can be used
+     * only after the integration is correctly started with startServer.
+     *
+     * @param msg      the message to forward. The ByteBuf in the message will be released.
+     * @param clientId the id of the sending integration.
+     * @throws IllegalStateException if the integration is not yet started
+     * @return
+     */
+    public RoutingResults internalPublish(MqttPublishMessage msg, final String clientId) {
+        final int messageID = msg.variableHeader().packetId();
+        if (!initialized) {
+            LOG.error("Moquette is not started, internal message cannot be published. CId: {}, messageId: {}", clientId,
+                      messageID);
+            throw new IllegalStateException("Can't publish on a integration is not yet started");
+        }
+        LOG.trace("Internal publishing message CId: {}, messageId: {}", clientId, messageID);
+        final RoutingResults routingResults = dispatcher.internalPublish(msg);
+        msg.payload().release();
+        return routingResults;
+    }
+
+    public void stopServer() {
+        LOG.info("Unbinding integration from the configured ports");
+        acceptor.close();
+        LOG.trace("Stopping MQTT protocol processor");
+        initialized = false;
+
+        // calling shutdown() does not actually stop tasks that are not cancelled,
+        // and SessionsRepository does not stop its tasks. Thus shutdownNow().
+        scheduler.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)